### PR TITLE
Add S&P/BR: M Rules + New Allegiance Headache

### LINKED ISSUE #1042: Add Shadow and Pain Rules
Assign this one to me.  I'll spruce 'em up when they release the scrolls in a couple of weeks.

### LINKED ISSUE #1043: Add Broken Realms: Morathi Rules
Assign to me for the release.  Rolling the dice that this won't be a nightmare amount of rules, lmao.

### PR DESCRIPTION
**Pull request by JohnSchramm**
Add Shadow & Pain rules (Fixes #1042).
Add Broken Realms: Morathi (Fixes #1043).
Modified allegiance abilities type to allow for selection.

Requires GUI modification to allow for allegiance selection within armies.
Requires warscroll import updates to adapt to new selection and type change.

### CODE REVIEW

File: src/army/daughters_of_khaine/units.ts [hunk 1]
```diff
@@ -471,6 +459,33 @@ export const Units: TUnits = [
...
+        name: `Blood of the Oracle`,
+        desc: `Roll a D6 each time this unit is affected by a spell or endless spell. On a 5+, ignore the effects on this model.`,
+        when: [HERO_PHASE],
+      },
+      {
+        name: `Gory Offering`,
+        desc: `If any enemy models were slain by this model in this phase, you can add 1 to the attacks characteristic of friendly Melusai units wholly within 12". Effect lasts until the end of the phase.`,
+        when: [COMBAT_PHASE],
+      },
+      TurnedCrystalEffect,
```
Comment (daviseford) [inline src/army/daughters_of_khaine/units.ts:475]: Good use of these effects. 

Btw I checked and when we share effects like this, it does reduce the final code bundle size, so the user benefits in multiple ways

File: src/army/daughters_of_khaine/units.ts [hunk 2]
```diff
@@ -471,6 +459,33 @@ export const Units: TUnits = [
...
+      {
+        name: `Wrath of the Scthborn`,
```
Comment (daviseford) [inline src/army/daughters_of_khaine/units.ts:477]: Is this really spelled `Scthborn`? If it is, that's wild 😆

File: src/army/idoneth_deepkin/abilities.ts [hunk 3]
```diff
@@ -11,31 +11,30 @@ const Abilities: TAbilities = [
   {
     name: `Isharann Rituals`,
     desc: `A maximum of one Isharann ritual can be performed in each hero phase.
-
+
```
Comment (daviseford) [inline src/army/idoneth_deepkin/abilities.ts:14]: Any reason for these tabs?

File: src/army/slaanesh/abilities.ts [hunk 4]
```diff
@@ -1,14 +1,19 @@
...
 // General Allegiance Abilities (always active regardless of army composition)
 const Abilities: TAbilities = [
   {
     name: `Feast of Depravities`,
     desc: `Each time a wound or mortal wound is inflicted on an enemy model by an attack or spell cast made by a friendly Slaanesh hero, and that enemy model is not slain by that wound, you receive 1 depravity point. In addition, every time a wound or mortal wound is allocated to a friendly Slaanesh Hero and not negated, and that friendly model is not slain by that wound or mortal would, you receive 1 depravity point.
-
+  
            Unit abilities and endless spell damage cannot generate depravity.`,
-    when: [DURING_GAME],
+    when: [WOUND_ALLOCATION_PHASE],
```
Comment (daviseford) [inline src/army/slaanesh/abilities.ts:16]: 👍

File: src/army/stormcast_eternals_stormkeep/index.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import Abilities from './abilities'
+import Allegiances from './allegiances'
+import Artifacts from './artifacts'
+import EndlessSpells from './endless_spells'
+import Spells from './spells'
+import Traits from './traits'
+import { AlliedUnits, Battalions, Units } from './units'
+
+// Copy paste this directory to kickstart a new army
```
Comment (daviseford) [inline src/army/stormcast_eternals_stormkeep/index.ts:9]: Remove boilerplate comments

File: src/army/stormcast_eternals_stormkeep/index.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+// Remember to add your army to:
+//      - meta/factions.ts
+//      - meta/army_list.ts
+
+// Be sure to rename this variable to `[FactionName]Army`
+const SampleArmy = {
```
Comment (daviseford) [inline src/army/stormcast_eternals_stormkeep/index.ts:15]:
```suggestion
const StormcastEternalsStormkeep = {
```

**Review by daviseford — CHANGES_REQUESTED** [review]
Looks awesome! I am checking on tests right now, as well as imports.

File: src/army/daughters_of_khaine/units.ts (hunk 2 above)
Comment (JohnSchramm) [reply src/army/daughters_of_khaine/units.ts:477]: Nope, typo lol.

Comment (exonian) [reply src/army/daughters_of_khaine/units.ts:477]: Scáthborn, I think

Comment (daviseford) [reply src/army/daughters_of_khaine/units.ts:477]: Just do a normal `a` character, please

File: src/army/daughters_of_khaine/units.ts [hunk 7]
```diff
@@ -5,15 +5,25 @@ import {
...
   MOVEMENT_PHASE,
   SAVES_PHASE,
   SHOOTING_PHASE,
   START_OF_HERO_PHASE,
   START_OF_ROUND,
+  TURN_FOUR_START_OF_ROUND,
   WOUND_ALLOCATION_PHASE,
 } from 'types/phases'
 
+const MorthaiEffect = {
```
Comment (exonian) [inline src/army/daughters_of_khaine/units.ts:20]:
```suggestion
const MorathiEffect = {
```

File: src/army/daughters_of_khaine/units.ts [hunk 8]
```diff
@@ -145,88 +167,60 @@ const BladedImpactEffect = {
...
-      {
-        name: `Enchanting Beauty`,
-        desc: `Subtract 1 from the hit rolls of attacks that target Morathi, High Oracle of Khaine.`,
-        when: [DURING_GAME],
-        tag: 'High Oracle',
+        name: `Commanding Presence`,
+        desc: `Subtract 1 from hit rolls targeting this model.`,
+        when: [SHOOTING_PHASE, COMBAT_PHASE],
       },
+      MorthaiEffect,
```
Comment (exonian) [inline src/army/daughters_of_khaine/units.ts:189]:
```suggestion
      MorathiEffect,
```

File: src/army/daughters_of_khaine/units.ts (hunk 2 above)
Comment (exonian) [reply src/army/daughters_of_khaine/units.ts:477]: Agreed 100%

**Comment by daviseford** [comment]
We are ending up with this pattern in Stormcast imports:

```
    +   Object {
    +     "severity": "ally-warn",
    +     "text": "Allied Prosecutors with Celestial Hammers can belong to Stormcast Eternals or Stormcast Eternals Stormkeep. Please add this unit manually.",
    +   },
    +   Object {
    +     "severity": "ally-warn",
    +     "text": "Allied Prosecutors with Stormcall Javelins can belong to Stormcast Eternals or Stormcast Eternals Stormkeep. Please add this unit manually.",
    +   },
```

Hmm....

File: src/army/stormcast_eternals_stormkeep/index.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  Abilities,
+  Allegiances,
+  AllegianceType: 'Stormkeeps',
```
Comment (exonian) [inline src/army/stormcast_eternals_stormkeep/index.ts:18]:
```suggestion
  AllegianceType: 'Stormhosts',
```

File: src/army/stormcast_eternals_stormkeep/index.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  AlliedUnits,
+  Artifacts,
+  Battalions,
+  EndlessSpells,
+  Spells,
+  Traits,
+  Units,
+}
+
+export default SampleArmy
```
Comment (exonian) [inline src/army/stormcast_eternals_stormkeep/index.ts:28]:
```suggestion
export default StormcastEternalsStormkeep
```

**Comment by daviseford** [comment]
Is `Morathi-Khaine` now distinct from `The Shadow Queen`? I remember we merged these units before - should we do that now, or are they now distinct?

**Comment by exonian** [comment]
> We are ending up with this pattern in Stormcast imports: 

How is this handled with the Legions of Nagash stuff?
*Is* it handled?

File: src/army/idoneth_deepkin/abilities.ts (hunk 3 above)
Comment (JohnSchramm) [reply src/army/idoneth_deepkin/abilities.ts:14]: Not sure.  Thought I pasted the original file back over to revert the abilities type change back out.  This file didn't change from a rules standpoint.

**Comment by daviseford** [comment]
> > We are ending up with this pattern in Stormcast imports:
> 
> How is this handled with the Legions of Nagash stuff?
> _Is_ it handled?

It's not handled very well, but luckily there isn't a HUGE overlap of LoN stuff - but yeah you'll definitely get some warnings when you import a generic LoN type army

**Comment by exonian** [comment]
> Is `Morathi-Khaine` now distinct from `The Shadow Queen`? I remember we merged these units before - should we do that now, or are they now distinct?

Two warscrolls and you always have both on the table now, as different models. So they need to be separate. Ideally side-effect-linked, but eh unless that got easier with the refactor it's not really a goer.

**Comment by daviseford** [comment]
Damn. I think after this PR gets in, we need to re-structure the idea of an army

File: src/army/daughters_of_khaine/units.ts [hunk 11]
```diff
@@ -145,88 +167,60 @@ const BladedImpactEffect = {
...
+const TurnedCrystalEffect = {
+  name: `Turned to Crystal`,
+  desc: `You can pick 1 enemy unit within 1" of this unit and roll D6 for each model in this unit. On each 3+ the target suffers 1 mortal wound.`,
+  when: [END_OF_COMBAT_PHASE],
+}
 
 // Unit Names
 export const Units: TUnits = [
   {
-    name: `Morathi, High Oracle of Khaine`,
```
[the remark this answers is not in the document]
Comment (JohnSchramm) [reply src/army/daughters_of_khaine/units.ts:155]: Sure.  I'll leave that one to the experts, lmao.

**Comment by JohnSchramm** [comment]
Yeah how Morathi works is a complete re-write.  She is two separate warscrolls/models on the board at the same time.

**Pull request merged**